client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala

```diff
@@ -1298,24 +1315,73 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
       shuffleId: Int,
       slotsToDestroy: WorkerResource): Unit = {
     val shuffleKey = Utils.makeShuffleKey(appUniqueId, shuffleId)
-    val parallelism = Math.min(Math.max(1, slotsToDestroy.size()), conf.clientRpcMaxParallelism)
-    ThreadUtils.parmap(
-      slotsToDestroy.asScala,
-      "DestroySlot",
-      parallelism) { case (workerInfo, (primaryLocations, replicaLocations)) =>
-      val destroy = DestroyWorkerSlots(
-        shuffleKey,
-        primaryLocations.asScala.map(_.getUniqueId).asJava,
-        replicaLocations.asScala.map(_.getUniqueId).asJava)
-      var res = requestWorkerDestroySlots(workerInfo.endpoint, destroy)
-      if (res.status != StatusCode.SUCCESS) {
-        logDebug(s"Request $destroy return ${res.status} for $shuffleKey, " +
-          s"will retry request destroy.")
-        res = requestWorkerDestroySlots(
-          workerInfo.endpoint,
-          DestroyWorkerSlots(shuffleKey, res.failedPrimarys, res.failedReplicas))
+
+    def retryDestroy(status: DestroyFutureWithStatus, currentTime: Long): Unit = {
+      status.retryTimes += 1
+      status.startTime = currentTime
+      status.future =
+        status.endpoint.ask[DestroyWorkerSlotsResponse](status.message)
+    }
+
+    val startTime = System.currentTimeMillis()
+    val futures = new util.LinkedList[DestroyFutureWithStatus]()
+    slotsToDestroy.asScala foreach { case (workerInfo, (primaryLocations, replicaLocations)) =>
+      val primaryIds = primaryLocations.asScala.map(_.getUniqueId).asJava
+      val replicaIds = replicaLocations.asScala.map(_.getUniqueId).asJava
+      val destroy = DestroyWorkerSlots(shuffleKey, primaryIds, replicaIds, mockDestroyFailure)
+      val future = workerInfo.endpoint.ask[DestroyWorkerSlotsResponse](destroy)
+      futures.add(DestroyFutureWithStatus(future, destroy, workerInfo.endpoint, 1, startTime))
+    }
+
+    val timeout = conf.rpcAskTimeout.duration.toMillis
+    var remainingTime = timeout * rpcMaxRetires
+    val delta = 50
+    while (remainingTime > 0 && !futures.isEmpty) {
+      val currentTime = System.currentTimeMillis()
+      val iter = futures.iterator()
+      while (iter.hasNext) {
+        val futureWithStatus = iter.next()
+        val message = futureWithStatus.message
+        val retryTimes = futureWithStatus.retryTimes
+        if (futureWithStatus.future.isCompleted) {
+          futureWithStatus.future.value.get match {
```

Comment:
If the future was not completed the returned value will be `None`

https://github.com/scala/scala/blob/v2.12.18/src/library/scala/concurrent/Future.scala#L182-L194

Reply:
```scala
futureWithStatus.future.value.foreach {
  case scala.util.Success(res) =>
    ...
  case scala.util.Failure(e) =>
    ...
  case _ =>
    ...
}